### PR TITLE
Added `MulAdd` and `MulAddAssign` traits

### PR DESCRIPTION
**Pull request by regexident**
Both `f32` and `f64` implement fused multiply-add, which computes `(self * a) + b` with only one rounding error. This produces a more accurate result with better performance than a separate multiplication operation followed by an add:

```rust
fn mul_add(self, a: f32, b: f32) -> f32[src]
```

It is however not possible to make use of this in a generic context by abstracting over a trait.

My concrete use-case is machine learning, [gradient descent](https://en.wikipedia.org/wiki/Gradient_descent) to be specific,  
where the core operation of updating the gradient could make use of `mul_add` for both its `weights: Vector` as well as its `bias: f32`:

```rust
struct Perceptron {
  weights: Vector,
  bias: f32,
}

impl MulAdd<f32, Self> for Vector {
  // ...
}

impl Perceptron {
  fn learn(&mut self, example: Vector, expected: f32, learning_rate: f32) {
    let alpha = self.error(example, expected, learning_rate);
    self.weights = example.mul_add(alpha, self.weights);
    self.bias = self.bias.mul_add(alpha, self.bias)
  }
}
```

(The actual impl of `Vector` would be generic over its value type: `Vector<T>`, thus requiring the trait.)

### CODE REVIEW

**Comment by cuviper** [comment]
Is this actually useful to you for integers?  Otherwise, you can just use [`Float::mul_add`](https://docs.rs/num-traits/0.2.2/num_traits/float/trait.Float.html#tymethod.mul_add).

File: src/ops/mul_add.rs [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    }
+}
+
+impl MulAdd<f64, f64> for f64 {
+    type Output = Self;
+
+    #[inline]
+    fn mul_add(self, a: Self, b: Self) -> Self::Output {
+        #![allow(unconditional_recursion)]
+        f64::mul_add(self, a, b)
```
Comment (cuviper) [inline src/ops/mul_add.rs]: I believe this `allow` is masking a real problem.  In a `#![no_std]` build, `f64` doesn't have most of its inherent methods, which means this call will resolve back to `MulAdd::mul_add` again -- thus the unconditional recursion.

**Comment by regexident** [comment]
> Is this actually useful to you for integers? Otherwise, you can just use `Float::mul_add`.

It's useful for writing code that makes use of the optimization if applicable, yet remains fully  generic. A particular use case is fixed-point arithmetic on `T: Real + MulAdd` as a convenient drop-in replacement for `T: Float + MulAdd`.

File: src/ops/mul_add.rs (hunk 1 above)
Comment (regexident) [reply src/ops/mul_add.rs]: Oh `#![no_std]` is what made the builds fail on CI, yet succeed on my local machine.
Adding the `#![allow(unconditional_recursion)]` felt fishy.

How would you propose solving this? Something like this?

```rust
if cfg!(feature == "std") {
    f64::mul_add(self, a, b)
} else {
    (self * a) + b
}
```

**Comment by vks** [comment]
> his produces a more accurate result with better performance than a separate multiplication operation followed by an add

Unfortunately, in my experience this is not necessarily true. For `f64` I don't see a performance difference on my machine, and for `f32` it is slower without `target-cpu=native` and faster with.

**Comment by regexident** [comment]
> Unfortunately, in my experience this is not necessarily true.

This does not invalidate the desire to have a way to generically express `fn mul_add` through a trait though, or does it?

**Comment by vks** [comment]
No, performance is not really related to the changes in this PR. What might be problematic is that `a * b + c` is less precise than `a.mul_add(b, c)` and will yield different results. I'm not sure how big of a problem that is in practice, but it should probably be documented.

Also see https://github.com/rust-lang/rust/pull/44805.

File: src/ops/mul_add.rs (hunk 1 above)
Comment (cuviper) [reply src/ops/mul_add.rs]: I think it would be better to *not* implement the trait at all for floats in `#![no_std]` mode, so we don't mislead about the rounding errors.  Either that, or do a manual implementation with appropriate rounding fixups, but I imagine that's non-trivial.

**Comment by regexident** [comment]
> No, performance is not really related to the changes in this PR. What might be problematic is that `a * b + c` is less precise than `a.mul_add(b, c)` and will yield different results. I'm not sure how big of a problem that is in practice, but it should probably be documented.

Good point, I'll gladly add a mention of this to the documentation. :)

Any further feedback? What needs to be done to proceed? Do we want to proceed?

**Comment by cuviper** [comment]
> > Is this actually useful to you for integers? Otherwise, you can just use `Float::mul_add`.
> 
> It's useful for writing code that makes use of the optimization if applicable, yet remains fully generic. A particular use case is fixed-point arithmetic on `T: Real + MulAdd` as a convenient drop-in replacement for `T: Float + MulAdd`.

Note there's [`Real::mul_add`](https://docs.rs/num-traits/0.2.2/num_traits/real/trait.Real.html#tymethod.mul_add) too.

Supporting the `no_std` fpa crate is a more compelling example though, since `Float` and `Real` are only available in `std` builds.

I still think we should *not* have a `(self * a) + b` fallback for `no_std` floats -- I'd rather not implement the trait for `no_std` floats at all if we can't meet the same rounding accuracy.  The performance is secondary, maybe not even worth bringing up.

**Comment by regexident** [comment]
I just moved the impls of `MulAdd`/`MulAddAssign` for `f32`/`f64` behind a feature guard with commit 28be885.

File: src/ops/mul_add.rs [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+            ($($t:ident)+) => {
+                $(
+                    {
+                        use core::$t;
+
+                        let m: $t = 12.0;
+                        let x: $t = 3.4;
+                        let b: $t = 5.6;
+
+                        let abs_difference = (m.mul_add(x, b) - (m*x + b)).abs();
```
Comment (cuviper) [inline src/ops/mul_add.rs]: I'm surprised this test didn't fail on `no_std`, but I think it's because `libtest` links to `libstd`, so the compiler still has the inherent impls of `f32`/`f64::mul_add` available.  Please change these tests to explicitly call `MulAdd::mul_add(m, x, b)`, and then this test will need a `cfg` gate.
